mangle `this` references when advantageous

laverdet opened the pull request:
By memoizing `this` in a variable it can be mangled just like other variables.

With this same technique we can also remove local calls to bind in most cases (unsafe mode only).

This yields significant savings in most OO code.

laverdet committed:
mangle `this` references when advantageous

By memoizing `this` in a variable it can be mangled just like other
variables.

With this same technique we can also remove local calls to bind in most
cases (unsafe mode only).

RGustBardon commented:
Fails when no source elements are present. When it comes to global code, it fails when a program is empty. When it comes to function code, it fails when it comes to

``` javascript
!function() {
};
```

but not when it comes to

``` javascript
!function() {
  42;
};
```

It is assumed (in the _EqualityExpression_ in line 913) that `body` has at least one element.

laverdet committed:
Fix ast_grab_this with empty bodies

function(){} would fail because `body` in this case is empty.

laverdet commented:
Aye. Not totally sure how to update the pull request, but 0e48c7db6f9579af34d07ed64bc5481b21ec75b2 fixes that easily.

laverdet commented:
There we go, added to pull request.

RGustBardon commented:
Results in more source characters in a scope where no mangling takes place (cf. `Scope.prototype.get_mangled` in `./lib/process.js`).

``` javascript

!function() {
  !function() {
    eval('');
  };
  void [typeof this, typeof this, typeof this, typeof this];
};

!function() {
  !function() {
    with ({});
  };
  void [typeof this, typeof this, typeof this, typeof this];
};
```

Adds a variable declaration in the global scope.

``` javascript
void [typeof this, typeof this, typeof this, typeof this];
```

michaelficarra commented:
Don't use `try`/`catch` for program control flow. Define `function needsContextCaching(body){}` or something.

Also, you're adding a `that` variable to the body (assigning `this` to it), but you aren't checking that the body doesn't already contain a `that` variable.

**edit:** I see you're inside a function already (`many_this`), and using the `try`/`catch` to escape from the walker function. That's a little trickier. I'll think about an improvement for that. If only we had block lambdas!

laverdet committed:
More fixes to `ast_grab_this`

- Failed in the case a function scope is not mangled (eval(), with{})
- Don't produce source that leaks globals
- Check for reserved mangle identifier

laverdet commented:
Thanks for the scrutiny :). This fixes the issue when no mangling takes place, and also takes care of the global leakage.

The try/catch was just an easy way to bail out of with_walkers() easily, but I replaced it with a function reassignment (read: monkey patch).

Re: `that` variable it's actually not `that`, it's `mangle_my_this_ref__`; `that` is a local shortcut. I assumed no one would use `mangle_my_this_ref__` in their source but went ahead and added a check just because.

michaelficarra commented:
Instead of bailing out when you see a variable named `mangle_my_this_ref__`, why not continually change it until it's no longer in scope?

laverdet commented:
idk it's a lot of legwork and you're a jerk if you're using that variable in your program.

michaelficarra commented:
It'd actually be shorter: `while(HOP(body.scope.refs, that)) that += '_';`

laverdet commented:
That will fail in the case where you have a `that` referenced in a nested function. It'll go into the nested function, update `that`, and then when it exits back out to the upper function it will have the wrong reference. You'd have to maintain them on the stack for each function you enter. It's only 3 more lines to handle that but I still think they're just a jerk if they use that reference.

RGustBardon commented:
`this` could be consolidated in the global scope by employing a closure (if it results in less source characters).

``` bash
uglifyjs -c <<ENDH
void [null, null, null, null, null, null, null, null, null, null];
with ({});
void [null, null, null, null, null, null, null, null, null, null];
ENDH
```

Line 973 of `./lib/process.js` (5523a29d55e9bd5fcb9143058f164a3b3c86fbb1) calls `unshift` on the `body`. This turns a Directive Prologue into an expendable sequence of _StringLiteral_ tokens.

``` javascript
!function() {
  'foo';
  'bar';
  void [typeof this, typeof this, typeof this, typeof this];
};
```

curiousdannii commented:
Please provide evidence that it produces shorter files once gzipped.

RGustBardon commented:
@curiousdannii It has been claimed otherwise in #228.

laverdet commented:
Interesting, I was not aware that this had been tried before. I did see gains in my code with and without gzip due to the bind() optimizations but for libraries which do not make use of the applicable bind() pattern it seems this would be harmful after gzip. That in mind, I may just take this walker and stick it in my own stack if common use of bind() in that way is uncommon with the community.
